inc/zoo/swar/SWAR.h

```diff
@@ -83,99 +88,153 @@ namespace detail {
+    constexpr SWAR operator^(SWAR o) const noexcept { return SWAR(m_v ^ o.m_v); }
 
-    constexpr bool operator==(SWAR o) { return m_v == o.m_v; }
-
-    constexpr T at(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        return filter & (m_v >> (Size * position));
+    constexpr T at(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        return filter & (m_v >> (NBits * position));
     }
 
-    constexpr SWAR clear(int position) {
-        constexpr auto filter = (T(1) << Size) - 1;
-        auto invertedMask = filter << (Size * position);
+    constexpr SWAR clear(int position) const noexcept {
+        constexpr auto filter = (T(1) << NBits) - 1;
+        auto invertedMask = filter << (NBits * position);
         auto mask = ~invertedMask;
         return SWAR(m_v & mask);
     }
 
-    constexpr int top() { return msb_index(m_v) / Size; }
-    constexpr int fastIndex() { return __builtin_ctzll(m_v) / Size; }
+    /// Notably, this is the SWAR lane index that contains the MSB.
+    constexpr int top() const noexcept { return msb_index(m_v) / NBits; }
+    /// Notably, this is the SWAR lane index that contains the lsb.
+    constexpr int lsbIndex() const noexcept { return __builtin_ctzll(m_v) / NBits; }
 
-    constexpr SWAR set(int index, int bit) {
-        return SWAR(m_v | (T(1) << (index * Size + bit)));
+    constexpr SWAR set(int index, int bit) const noexcept {
+        return SWAR(m_v | (T(1) << (index * NBits + bit)));
     }
 
-    constexpr operator bool() const { return m_v; }
-protected:
     T m_v;
 };
 
-template<int Size, typename T>
-struct BooleanSWAR: SWAR<Size, T> {
-    constexpr BooleanSWAR(T v): SWAR<Size, T>(v) {}
+/// Defining operator== on base SWAR types is entirely too error prone. Force a verbose invocation.
+template<int NBits, typename T = uint64_t> constexpr auto horizontalEquality(SWAR<NBits, T> left, SWAR<NBits, T> right) {
+  return left.m_v == right.m_v;
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolate(T pattern) {
+  return pattern & ((T(1)<<NBits)-1);
+}
+
+template<typename T = uint64_t>
+constexpr auto clearLSB(T v) {
+  return v & (v - 1);
+}
+
+template<typename T = uint64_t>
+constexpr auto isolateLSB(T v) {
+  return v & ~clearLSB(v);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto maskLowBits() {
+  return (T(1)<<(NBits-1)) | ((T(1)<<(NBits-1))-1);
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto clearLSBits(T v) {
+  return v &(~(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v))));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto isolateLSBits(T v) {
+  return v &(maskLowBits<NBits>() << metaLogFloor(isolateLSB<T>(v)));
+}
+
+template<int NBits, typename T = uint64_t>
+constexpr auto broadcast(SWAR<NBits, T> v) {
+  constexpr T Ones = makeBitmask<NBits, T>(SWAR<NBits, T>{1}.value());
+  return SWAR<NBits, T>(T(v) * Ones);
```

Comment:
Please make a note this assumes the input has **only the desired value in the least significant "lane"**

Reply:
Done

Reply:
Using "lanes" the clarity  improved a lot